# Data Collection

## Import packages

In [1]:
import pandas as pd
import numpy as np

## Import csv files

A .names file accompanied the csv files which provided descriptions for all attributes of the data including the original dtypes and feature descriptions.

In [69]:
assess = pd.read_csv('assessments.csv', na_values = '?', low_memory = False)
courses = pd.read_csv('courses.csv', na_values = '?', low_memory = False)
stassess = pd.read_csv('studentAssessment.csv', na_values = '?', low_memory = False)
stinfo = pd.read_csv('studentInfo.csv', na_values = '?', low_memory = False)
streg = pd.read_csv('studentRegistration.csv', na_values = '?', low_memory = False)
stvle = pd.read_csv('studentVle.csv', na_values = '?', low_memory = False)
vle = pd.read_csv('vle.csv', na_values = '?', low_memory = False)

# Preliminary data exploration

## Baseline Data Info

In [70]:
assess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    object 
 1   code_presentation  206 non-null    object 
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    object 
 4   date               195 non-null    float64
 5   weight             206 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 9.8+ KB


In [71]:
assess.shape

(206, 6)

In [72]:
assess.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

In [73]:
courses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   code_module                 22 non-null     object
 1   code_presentation           22 non-null     object
 2   module_presentation_length  22 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 660.0+ bytes


In [74]:
courses.shape

(22, 3)

In [75]:
courses.head()

code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [76]:
stassess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id_assessment   173912 non-null  int64  
 1   id_student      173912 non-null  int64  
 2   date_submitted  173912 non-null  int64  
 3   is_banked       173912 non-null  int64  
 4   score           173739 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 6.6 MB


In [77]:
stassess.shape

(173912, 5)

In [78]:
stassess.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [79]:
stinfo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   code_module           32593 non-null  object
 1   code_presentation     32593 non-null  object
 2   id_student            32593 non-null  int64 
 3   gender                32593 non-null  object
 4   region                32593 non-null  object
 5   highest_education     32593 non-null  object
 6   imd_band              31482 non-null  object
 7   age_band              32593 non-null  object
 8   num_of_prev_attempts  32593 non-null  int64 
 9   studied_credits       32593 non-null  int64 
 10  disability            32593 non-null  object
 11  final_result          32593 non-null  object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


In [80]:
stinfo.shape

(32593, 12)

In [81]:
stinfo.head()

code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         AAA             2013J       28400      F              Scotland   
2         AAA             2013J       30268      F  North Western Region   
3         AAA             2013J       31604      F     South East Region   
4         AAA             2013J       32885      F  West Midlands Region   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1       HE Qualification   20-30%    35-55                     0   
2  A Level or Equivalent   30-40%    35-55                     0   
3  A Level or Equivalent   50-60%    35-55                     0   
4     Lower Than A Level   50-60%     0-35                     0   

   studied_credits disability final_result  
0              240          N         Pass  
1               60          N         Pass  
2               60          Y    Withdrawn  
3               60          N         Pass  
4               60          N         Pass

In [82]:
streg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   code_module          32593 non-null  object 
 1   code_presentation    32593 non-null  object 
 2   id_student           32593 non-null  int64  
 3   date_registration    32548 non-null  float64
 4   date_unregistration  10072 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 1.2+ MB


In [83]:
streg.shape

(32593, 5)

In [84]:
streg.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

In [85]:
stvle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10655280 entries, 0 to 10655279
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   code_module        object
 1   code_presentation  object
 2   id_student         int64 
 3   id_site            int64 
 4   date               int64 
 5   sum_click          int64 
dtypes: int64(4), object(2)
memory usage: 487.8+ MB


In [86]:
stvle.shape

(10655280, 6)

In [87]:
stvle.head()

code_module code_presentation  id_student  id_site  date  sum_click
0         AAA             2013J       28400   546652   -10          4
1         AAA             2013J       28400   546652   -10          1
2         AAA             2013J       28400   546652   -10          1
3         AAA             2013J       28400   546614   -10         11
4         AAA             2013J       28400   546714   -10          1

In [88]:
vle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6364 entries, 0 to 6363
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_site            6364 non-null   int64  
 1   code_module        6364 non-null   object 
 2   code_presentation  6364 non-null   object 
 3   activity_type      6364 non-null   object 
 4   week_from          1121 non-null   float64
 5   week_to            1121 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 298.4+ KB


In [89]:
vle.shape

(6364, 6)

In [90]:
vle.head()

id_site code_module code_presentation activity_type  week_from  week_to
0   546943         AAA             2013J      resource        NaN      NaN
1   546712         AAA             2013J     oucontent        NaN      NaN
2   546998         AAA             2013J      resource        NaN      NaN
3   546888         AAA             2013J           url        NaN      NaN
4   547035         AAA             2013J      resource        NaN      NaN

## Helper elements

I created helper elements that would be beneficial for some later looping code including a list of all the dataframes and assigning each df with a name

In [91]:
ls_all_docs = [assess, courses, stassess, stinfo, streg, stvle, vle]
assess.name = 'assess'
courses.name = 'courses'
stassess.name = 'stassess'
stinfo.name = 'stinfo'
streg.name = 'streg'
stvle.name = 'stvle'
vle.name = 'vle'

## Checking for Duplicates

In [92]:
for i in ls_all_docs:
    colu = [i.columns]
    for c in colu:
        if i[c].duplicated().any() is True:
            print(f'{i.name}[{c}] has duplicates')
        else:
            print(f'{i.name} has no duplicates')

assess has no duplicates
courses has no duplicates
stassess has no duplicates
stinfo has no duplicates
streg has no duplicates
stvle has no duplicates
vle has no duplicates


# Joining Data

## Finding merge features

I needed to identify the specific data features that were in common for each df. The first for loop required cross referencing multiple posts and sources (including my notes) to figure out the specific syntax.

In [93]:
similar_headers = []

for i, df_outer in enumerate(ls_all_docs):
    for j, df_inner in enumerate(ls_all_docs):
        if i == j:
            continue
        for c in df_outer.columns:
            if c in df_inner.columns and c not in similar_headers:
                similar_headers.append(c)

In [94]:
similar_headers

['code_module',
 'code_presentation',
 'id_assessment',
 'date',
 'id_student',
 'id_site']

In [96]:
for h in similar_headers:
    for d in ls_all_docs:
        ls = list(d.columns)
        if h in ls:
            print(f"{d.name} has {h}")

assess has code_module
courses has code_module
stinfo has code_module
streg has code_module
stvle has code_module
vle has code_module
assess has code_presentation
courses has code_presentation
stinfo has code_presentation
streg has code_presentation
stvle has code_presentation
vle has code_presentation
assess has id_assessment
stassess has id_assessment
assess has date
stvle has date
stassess has id_student
stinfo has id_student
streg has id_student
stvle has id_student
stvle has id_site
vle has id_site


With my common features identified, I then physically constructed a table to visualize which features overlap. To condense the number of repeat columns between merges, I chose to list all common features between data frames that were possible at a time.

## Joining attempt 1

All merges follow a similar pattern where all common features were used to merge onto. Additionally, df that were assumed to inform another df were merged first. The head method was added later to highlight something that I was not able to catch prior to merging.

Each merge was also checked to ensure there were no missing values. This will actually be identified as incorrect exemplified with 'sacri'.

as_co = assess.merge(courses, on = ['code_module', 'code_presentation'])
as_co.isna().sum()

as_co.info()

as_co.head()

as_co.shape

sta_ac = stassess.merge(as_co, on = 'id_assessment')
sta_ac.isna().sum()

sta_ac.info()

sta_ac.head()

sta_ac.shape

in_re = stinfo.merge(streg, on = ['code_module', 'code_presentation', 'id_student'])
in_re.isna().sum()

in_re.info()

in_re.head()

in_re.shape

sv_vl = stvle.merge(vle, on = ['id_site', 'code_module', 'code_presentation'])
sv_vl.isna().sum()

sv_vl.info()

sv_vl.head()

sv_vl.shape

sacri = sta_ac.merge(in_re, on = ['code_module', 'code_presentation', 'id_student'])
sacri.isna().sum()

sacri.info()

sacri.head()

sacri.shape

## Join error

The following merge produced error that indicated that the date feature in sacri is considered an object whereas the date column in sv_vl is considered an Int. Looking at the .names file accompanying document, the feature should be populated with integers indicating the time passed since the start of the course.

#merged_df = sv_vl.merge(sacri, on = ['code_module', 'code_presentation', 'date', 'id_student'])
#merged_df.isna().sum()

## Diagnosis

I first needed to identify what characters were being considered as objects rather than int's.

values_not_int = []
values_int = []

for index, row in sacri.iterrows():
    if row['date'] is not int:
        values_not_int.append(row['date'])
        print(f'{row["date"]}: {type(row["date"])}')
    else:
        values_int.append(row['date'])

values_int

The loop indicated that all values were considered strings. So I needed to convert all values to an integer.

The following code produced an error since '?' cannot be converted to int dtype.

#for index, row in sacri.iterrows():
#    row['date'] = int(row['date'])

sacri[sacri.eq('?').any(1)]

Seeing that date_unregistration had '?', I decided to check all other merge stages.

#in_re[in_re.eq('?').any(1)]

#sta_ac[sta_ac.eq('?').any(1)]

#as_co[as_co.eq('?').any(1)]

#sv_vl[sv_vl.eq('?').any(1)]

All merges have missing data in the form of '?' which is why the initial merges did not show any missing data.

## Missing original data

### Assess

Assess lists out the assignments associated with the respective modules and presentations.

assess[assess.eq('?').any(1)]

Since the missing data is the date the final submission date (int) for each assessment, we cannot drop those rows without damaging other df during merging. Since no supplemental data is available to determine the date, these dates will need to be imputed.

Assuming that the last day to turn in any assignment is the end of the course, each value was imputed with the maximum date within the feature.

assess['date'].replace('?', np.nan, inplace = True)
assess['date'] = pd.to_numeric(assess['date'], errors = 'coerce')
for c in assess.columns:
    assess_max = assess['date'].max()
    assess['date'].replace('?', assess_max, inplace = True)
#assess['date'] = assess['date'].astype('int64')

assess['date'].dtypes

assess.info()

assess.isnull().sum()

assess[assess['date'].isnull()]

assess.info()

courses[courses.eq('?').any(1)]

stassess[stassess.eq('?').any(1)]

stinfo[stinfo.eq('?').any(1)]

streg[streg.eq('?').any(1)]

stvle[stvle.eq('?').any(1)]

vle[vle.eq('?').any(1)]

print(sacri['date'].isna().sum()/len(sacri['date']))

I am choosing to dropna all rows that have had a '?' converted to NaN since the amount of missing data is less than 2%

sacri.dropna(subset = ['date'], inplace = True)
sacri.shape

## Joining attempt 2

After speaking with my mentor, I realized that there is a more streamline way to replace the '?' with NaN objects. I used the pd.read_csv() argument na_values = '?' to load the files with the NaN values. The following section will try to parse out what data I need to impute or manage prior to merging.

In [104]:
for i in ls_all_docs:
    print(i.name)
    print(i.isna().sum())
    print(i.shape)
    print()

assess
code_module           0
code_presentation     0
id_assessment         0
assessment_type       0
date                 11
weight                0
dtype: int64
(206, 6)

courses
code_module                   0
code_presentation             0
module_presentation_length    0
dtype: int64
(22, 3)

stassess
id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64
(173912, 5)

stinfo
code_module                0
code_presentation          0
id_student                 0
gender                     0
region                     0
highest_education          0
imd_band                1111
age_band                   0
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64
(32593, 12)

streg
code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64
(32593, 5)

### Assessment

Since the missing data is the date the final submission date (int) for each assessment, we cannot drop those rows without damaging other df during merging. Since no supplemental data is available to determine the date, these dates will need to be imputed.

Assuming that the last day to turn in any assignment is the end of the course, each value was imputed with the maximum date within the feature.

In [107]:
assess['date'] = assess['date'].fillna(assess['date'].max())

In [108]:
assess.isna().sum()

code_module          0
code_presentation    0
id_assessment        0
assessment_type      0
date                 0
weight               0
dtype: int64

### Student Assessment

In [112]:
ranges = [0, 40, 50, 60, 70, 80, 90, 100]
stassess.groupby(pd.cut(stassess['score'], ranges)).count()

id_assessment  id_student  date_submitted  is_banked  score
score                                                                 
(0, 40]             9215        9215            9215       9215   9215
(40, 50]            7401        7401            7401       7401   7401
(50, 60]           16561       16561           16561      16561  16561
(60, 70]           23920       23920           23920      23920  23920
(70, 80]           40752       40752           40752      40752  40752
(80, 90]           36330       36330           36330      36330  36330
(90, 100]          39231       39231           39231      39231  39231

Missing values imply that the grade was not able to be reported for each of the 173 student assignments. Since the scores are being investigated as dependent variables, their presence is crucial to the study. Dropping the 173 student assignments will minimally (0.09% of the sample population) impact the study.

In [115]:
stassess_drop = stassess.dropna()

In [116]:
stassess_drop.isna().sum()

id_assessment     0
id_student        0
date_submitted    0
is_banked         0
score             0
dtype: int64